pybpodapi/bpod/emulator/emulator.py

```diff
@@ -0,0 +1,668 @@
+                        this_global_timer_idx]
+                self._global_timer_end[
+                    this_global_timer_idx] = \
+                    self._global_timer_start[
+                        this_global_timer_idx] + \
+                    self._state_machine.global_timers.timers[
+                        this_global_timer_idx]
+                self._global_timers_triggered[this_global_timer_idx] = True
+            logger.debug('Global timer end time set.')
+
+    def set_softcode(self, output_code):
+        self._softcode = self._output_value(
+            output_code=output_code)
+        logger.debug(f'Softcode set to: {self._softcode}')
+
+    def reset_current_state(self):
+        self._assert_state_machine()
+        self._current_state = self._state_machine.current_state
+        logger.debug(f'Current state reset to: {self._current_state}')
+
+    def reset_global_counter_counts(self):
+        this_global_counter = self._output_value(
+            GLOBAL_COUNTER_RESET_OUTPUT_ACTION)
+        if this_global_counter != 0:
+            self._global_counter_counts[this_global_counter - 1] = 0
+            logger.debug('Global counter counts reset.')
+
+    def cancel_global_timers(self):
+        this_global_timer = self._output_value(
+            GLOBAL_TIMER_CANCEL_OUTPUT_ACTION)
+        if this_global_timer != 0:
+            self._global_timers_active[this_global_timer - 1] = False
+            logger.debug('Global timers cancelled.')
+
+    def initialize(self):
+        self._assert_state_machine()
+        self._n_events = 0
+        self._current_state = 0
+        self._global_timer_start = [0] * self.hardware.n_global_timers
+        self._global_timer_end = [0] * self.hardware.n_global_timers
+        self._global_timers_triggered = [
+            False] * self.hardware.n_global_timers
+        self._global_timers_active = [False] * self.hardware.n_global_timers
+        self._global_counter_counts = [0] * self.hardware.n_global_counters
+        self._condition_channels = [0] * self.hardware.n_conditions
+        self._condition_values = [0] * self.hardware.n_conditions
+        self._timestamps = [0] * N_TIMESTAMPS
+        self._meaningful_timer = []
+        for tup in list(zip(
+                self._state_machine.state_timer_matrix,
+                range(self._state_machine.total_states_added))):
+            self._meaningful_timer.append(tup[0] != tup[1])
+        self.current_time = time.time()
+        self._matrix_start_time = self.current_time
+        self._state_start_time = self.current_time
+        self._softcode = self._output_value(SOFTCODE_OUTPUT_ACTION)
+        # Set global timer end-time (if triggered in first state)
+        self.set_global_timer_end_time(set_global_channel_timer=True)
+
+    def run(self):
+        self._assert_state_machine()
+        virtual_current_events = []
+        # MATLAB EQUIVALENT:
+        # virtual_current_events = [0] * N_VIRTUAL_CURRENT_EVENTS
```

Comment:
I think it would be more wise to keep it simple and pythonic as possible for this PR. And maybe consider using numpy in this part but also elsewhere as part of a separate PR.